Services/DefaultUserService.cs

```diff
@@ -63,10 +73,64 @@ public async Task UpdateUserPassword(User update)
         {
             var user = FindByUsername(update.Username);
             if (user == null)
-                throw new Exception("User not found");
+                throw new InvalidOperationException("User not found");
 
             user.Password = Encrypter.EncryptSHA256(update.Password);
             await _context.SaveChangesAsync();
         }
+
+        public Contact? FindContactByUsername(string username)
+        {
+            return _context.Contacts
+                .SingleOrDefault(c => c.UserUsername == username);
+        }
+
+        public async Task AddContact(Contact contact)
+        {
+            var existing = FindContactByUsername(contact.UserUsername);
+            if (existing != null)
+            {
+                throw new InvalidOperationException("User contact already exists");
+            }
+            _context.Contacts.Add(contact);
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task UpdateContact(Contact update)
+        {
+            var existing = FindContactByUsername(update.UserUsername);
+            if (existing == null)
+            {
+                throw new InvalidOperationException("User contact doesn't exist");
+            }
+            existing.Phone = update.Phone;
+            existing.Address = update.Address;
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task DeleteContactByUsername(string username)
+        {
+            var existing = FindContactByUsername(username);
+
+            if (existing == null)
+            {
+                throw new InvalidOperationException("User contact doesn't exist");
+            }
+            _context.Contacts.Remove(existing);
+            await _context.SaveChangesAsync();
+        }
+
+        public async Task AddOrUpdateContactIfExists(Contact contact)
```

Comment:
Should be 2 methods: `AddContact()` vs `UpdateContact()`. **One method** should take only **one responsibility** at a time (following by **S**OLID principles)